File: amr/localization/motion_model.py
"""Thrun odometry motion model (sample_motion_model_odometry).

Decomposes the robot-frame relative odometry into rot1 / trans / rot2 and
applies independent per-particle Gaussian noise, vectorized over all particles.
See Probabilistic Robotics (Thrun et al.) Table 5.6.
"""
from __future__ import annotations

import math
from typing import Sequence

import numpy as np

from amr.core.geometry import wrap_angle, wrap_angles
from amr.core.types import Pose2D


def sample_motion(particles: np.ndarray, odom_delta: Pose2D,
                  alphas: Sequence[float], rng: np.random.Generator) -> np.ndarray:
    """Thrun odometry motion model applied to (N,3) particles; returns new (N,3).

    ``odom_delta`` is the robot-frame relative motion (dx, dy, dth) since the
    last update. ``alphas`` are the four noise coefficients (a1..a4).
    """
    particles = np.asarray(particles, dtype=float)
    p = particles.copy()
    n = p.shape[0]

    a1, a2, a3, a4 = float(alphas[0]), float(alphas[1]), float(alphas[2]), float(alphas[3])

    dx, dy, dth = float(odom_delta.x), float(odom_delta.y), float(odom_delta.theta)

    trans = math.hypot(dx, dy)
    rot1 = math.atan2(dy, dx) if trans > 1e-4 else 0.0
    rot2 = wrap_angle(dth - rot1)

    abs_rot1 = abs(rot1)
    abs_rot2 = abs(rot2)

    std_rot1 = a1 * abs_rot1 + a2 * trans
    std_trans = a3 * trans + a4 * (abs_rot1 + abs_rot2)
    std_rot2 = a1 * abs_rot2 + a2 * trans

    rot1_s = rot1 + rng.normal(0.0, std_rot1, n)
    trans_s = trans + rng.normal(0.0, std_trans, n)
    rot2_s = rot2 + rng.normal(0.0, std_rot2, n)

    th = p[:, 2]
    p[:, 0] = p[:, 0] + trans_s * np.cos(th + rot1_s)
    p[:, 1] = p[:, 1] + trans_s * np.sin(th + rot1_s)
    p[:, 2] = wrap_angles(th + rot1_s + rot2_s)

    return p

File: src/amr_localization/src/motion_model.cpp
// SPDX-License-Identifier: Apache-2.0
#include "amr_localization/motion_model.hpp"

#include <cmath>

#include "amr_core/geometry.hpp"

namespace amr_localization {

std::vector<amr_core::Pose2D> sample_motion(
    const std::vector<amr_core::Pose2D>& particles,
    const amr_core::Pose2D& odom_delta, const std::vector<double>& alphas,
    std::mt19937& rng) {
  const double a1 = alphas[0];
  const double a2 = alphas[1];
  const double a3 = alphas[2];
  const double a4 = alphas[3];

  const double dx = odom_delta.x;
  const double dy = odom_delta.y;
  const double dth = odom_delta.theta;

  const double trans = std::hypot(dx, dy);
  const double rot1 = (trans > 1e-4) ? std::atan2(dy, dx) : 0.0;
  const double rot2 = amr_core::wrap_angle(dth - rot1);

  const double abs_rot1 = std::abs(rot1);
  const double abs_rot2 = std::abs(rot2);

  const double std_rot1 = a1 * abs_rot1 + a2 * trans;
  const double std_trans = a3 * trans + a4 * (abs_rot1 + abs_rot2);
  const double std_rot2 = a1 * abs_rot2 + a2 * trans;

  // Per-particle independent draws (one rot1/trans/rot2 sample each), matching
  // numpy's rng.normal(mean, std, n). A zero std yields the mean exactly.
  std::normal_distribution<double> n_rot1(0.0, std_rot1);
  std::normal_distribution<double> n_trans(0.0, std_trans);
  std::normal_distribution<double> n_rot2(0.0, std_rot2);

  std::vector<amr_core::Pose2D> out(particles.size());
  for (std::size_t i = 0; i < particles.size(); ++i) {
    const double rot1_s = rot1 + n_rot1(rng);
    const double trans_s = trans + n_trans(rng);
    const double rot2_s = rot2 + n_rot2(rng);

    const double th = particles[i].theta;
    out[i].x = particles[i].x + trans_s * std::cos(th + rot1_s);
    out[i].y = particles[i].y + trans_s * std::sin(th + rot1_s);
    out[i].theta = amr_core::wrap_angle(th + rot1_s + rot2_s);
  }
  return out;
}

}  // namespace amr_localization
